File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.ans = -1e9
        def solve(root):
            if not root: return 0
            left = max(solve(root.left), 0)
            right = max(solve(root.right), 0)
            tmp = root.val + left+right
            self.ans = max(self.ans, tmp)
            return max(root.val + left, root.val+right)
        solve(root)
        return self.ans

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int pathSum(TreeNode root, int[] arr){
        if(root == null) return 0;

        int leftSum = Math.max(0,pathSum(root.left, arr));
        int rightSum = Math.max(0,pathSum(root.right, arr));

        arr[0] = Math.max(arr[0], rightSum + leftSum + root.val );

        return root.val + Math.max(leftSum,rightSum);
    }
    public int maxPathSum(TreeNode root) {
        if(root==null) return 0;
        int[] arr = {root.val};
        pathSum(root,arr);
        return arr[0];
    }
}